### PR TITLE
stm32: add support for i2c read

### PR DESCRIPTION
**Pull request by Arksine**
This implementation is a compromise of "method 1" outlined in the STM32F1 reference manual (RM0041, Page 579), where critical sections are protected by disabling IRQs.  Method 2 is indicated for polling, however it seems unnecessary and overly complex with code paths for 1, 2 and 3+ bytes read.

Tested working on both the STM32F1 (BTT SKR Mini E3) and STM32F4 (SKR Pro).

Signed-off-by:  Eric Callahan <arksine.code@gmail.com>

### CODE REVIEW

**Comment by dianlight** [comment]
my2cents

Can you check if your implementation work for 0 len write before read? 

I don't have the board so I can't test whether to start writing and then start reading without writing can give problems.

As a reference #2795 and PR #2803

**Comment by Arksine** [comment]
I suspect as it currently is it wouldn't work because the write sequence would be started.  It should be simple enough to add.  I'll post up a patch then test it when I get a chance.

File: src/stm32/i2c.c [hunk 1]
```diff
@@ -81,13 +82,20 @@ i2c_wait(I2C_TypeDef *i2c, uint32_t set, uint32_t clear, uint32_t timeout)
...
+i2c_start(I2C_TypeDef *i2c, uint8_t addr, uint8_t xfer_len,
+          uint32_t timeout)
 {
     i2c->CR1 = I2C_CR1_START | I2C_CR1_PE;
     i2c_wait(i2c, I2C_SR1_SB, 0, timeout);
     i2c->DR = addr;
+    if (addr & 0x01)
+        i2c->CR1 |= I2C_CR1_ACK;
     i2c_wait(i2c, I2C_SR1_ADDR, 0, timeout);
+    irq_disable();
```
Comment (KevinOConnor) [inline src/stm32/i2c.c:94]: For intermediate code like this, it's better to use `irqstatus_t flag = irq_save(); ...; irq_restore(flag);` - that way the code does not enable interrupts if an i2c transfer was actually invoked while irqs were disabled.

File: src/stm32/i2c.c [hunk 2]
```diff
@@ -99,6 +107,18 @@ i2c_send_byte(I2C_TypeDef *i2c, uint8_t b, uint32_t timeout)
     i2c_wait(i2c, I2C_SR1_TXE, 0, timeout);
 }
 
+static void
+i2c_read_byte(I2C_TypeDef *i2c, uint8_t* read, uint32_t timeout,
+              uint8_t remaining)
+{
+    i2c_wait(i2c, I2C_SR1_RXNE, 0, timeout);
+    irq_disable();
+    *read = i2c->DR;
```
Comment (KevinOConnor) [inline src/stm32/i2c.c:116]: Very minor, comment - gcc often does a poor job generating code for pass-by-reference parameters.  Might be better to change this function to return a uint8_t.

**Comment by Arksine** [comment]
Thanks Kevin.  I implemented the suggested changes.  FWIW, the AVR i2c code also passes the `read` pointer to "i2c_receive_byte", if you would like I could change it to return a uint8_t and submit a patch for it as well.

**Comment by KevinOConnor** [comment]
Thanks.

The passing by reference is a very minor comment, so not a big deal either way.

Cheers,
-Kevin

**Pull request merged**